src/ModelContextProtocol/Protocol/Transport/StreamableHttpPostTransport.cs

```diff
@@ -0,0 +1,134 @@
+    }
+
+    public async Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
+    {
+        await _sseWriter.SendMessageAsync(message, cancellationToken).ConfigureAwait(false);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        await _sseWriter.DisposeAsync().ConfigureAwait(false);
+    }
+
+    private async IAsyncEnumerable<SseItem<JsonRpcMessage?>> StopOnFinalResponseFilter(IAsyncEnumerable<SseItem<JsonRpcMessage?>> messages, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        await foreach (var message in messages.WithCancellation(cancellationToken))
+        {
+            yield return message;
+
+            if (message.Data is JsonRpcResponse response)
+            {
+                if (_pendingRequests.Remove(response.Id) && _pendingRequests.Count == 0)
+                {
+                    // Complete the SSE response stream now that all pending requests have been processed.
+                    break;
+                }
+            }
+        }
+    }
+
+    private async ValueTask OnPostBodyReceivedAsync(PipeReader streamableHttpRequestBody, CancellationToken cancellationToken)
+    {
+        if (!await IsJsonArrayAsync(streamableHttpRequestBody, cancellationToken).ConfigureAwait(false))
+        {
+            var message = await JsonSerializer.DeserializeAsync(streamableHttpRequestBody.AsStream(), McpJsonUtilities.JsonContext.Default.JsonRpcMessage, cancellationToken).ConfigureAwait(false);
+            await OnMessageReceivedAsync(message, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            // Batched JSON-RPC message
+            var messages = JsonSerializer.DeserializeAsyncEnumerable(streamableHttpRequestBody.AsStream(), McpJsonUtilities.JsonContext.Default.JsonRpcMessage, cancellationToken).ConfigureAwait(false);
+            await foreach (var message in messages.WithCancellation(cancellationToken))
+            {
+                await OnMessageReceivedAsync(message, cancellationToken).ConfigureAwait(false);
+            }
+        }
+    }
+
+    private async ValueTask OnMessageReceivedAsync(JsonRpcMessage? message, CancellationToken cancellationToken)
+    {
+        if (message is null)
+        {
+            throw new InvalidOperationException("Received invalid null message.");
+        }
+
+        if (message is JsonRpcRequest request)
+        {
+            _pendingRequests.Add(request.Id);
+        }
+
+        message.RelatedTransport = this;
+
+        // Really an assertion. This doesn't get called when incomingChannel is null for GET requests.
+        Throw.IfNull(incomingChannel);
+        await incomingChannel.WriteAsync(message, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async ValueTask<bool> IsJsonArrayAsync(PipeReader requestBody, CancellationToken cancellationToken)
+    {
+        // REVIEW: Should we bother trimming whitespace before checking for '['?
```

Comment:
Pros / cons? Is it compliant with the spec to have whitespace there?

Reply:
I don't think the spec says anything about whitespace, but it if it does, it'd probably want us to be tolerant. Still, I'd lean against trimming until we hear about compatibility complaints for convenience of implementation.